compiler/resolution/preFold.cpp

```diff
@@ -742,12 +739,10 @@ static Expr* preFoldPrimOp(CallExpr* call) {
     }
 
   } else if (call->isPrimitive(PRIM_NUM_FIELDS)) {
-    SymExpr*       se1        = toSymExpr(call->get(1));
-    AggregateType* classType  = toAggregateType(se1->symbol()->type);
+    Type*          t          = canonicalClassType(call->get(1)->getValType());
```

Comment:
It seems a shame to have all these pairs of `Type t = canonicalClassType(...); classType = toAggregateType(t);` pairs.  Should canonicalClassType return an AggregateType?  Or should there be a version that always gives you an AggregateType?

Reply:
I think it's fine to have the pairs, in the same way that getValType and toAggregateType are different operations. Nonetheless we could certainly add a canonicalAggregateType or something.